### ISSUE TITLE
Allow to select between luks and luks2 when creating encrypted partition

### ISSUE THREAD

**Issue by Bleuzen** [calamares/calamares#1643]
**Is your feature request related to a problem? Please describe.**
Yes, I want to use luks2 for encryption. This is currently not possible as any encrypted partition created with Calamares ends up being luks1.

**Describe the solution you'd like**
Add a simple checkbox in the manual partition creation dialog for "Use luks2":
![image](https://user-images.githubusercontent.com/12885163/108502204-bae13a80-72b2-11eb-9195-3654dec3dcdb.png)

Also display a warning message like this when the user first clicks this checkbox:

**"WARNING: Luks2 is currently not fully supported by grub. Only use this option for pure data partitions or if you are using unencrypted /boot, otherwise your system won't boot!"**

**Describe alternatives you've considered**
For implementation of this feature:
I don't think there would be a much better alternative to implement this.

To achieve a setup with luks2 partitions now:
The only option is to not use Calamares.

**Additional context**
Nobody should be forced to use an old technology if there is no reason to.
Users should be able to enjoy the benefits of luks2 on any partition which does not hold encrypted /boot.
This is useful for example:
- for data partitions (like /home and other disks)
- when using unencrypted /boot
- when using separate encrypted /boot and encrypted root

Therefore it should be possible to select which partitions really still need luks1 and which should be luks2.

**Comment by philmmanjaro** [comment]
See https://github.com/calamares/calamares/issues/1099

**Comment by Bleuzen** [comment]
Hi @philmmanjaro , the issue you linked is about the changed / current default of cryptsetup to use luks2.
However, my feature request here has nothing to do with the default.
This is about giving the users an option to manually select the luks version.

**Comment by adriaandegroot** [comment]
There are three parts to this request:
- luks2 needs to work **at all**; the issue Phil refers to discusses that and it's not clear that all the parts are in place to do that yet; this will also depend on the distro and the other parts of that stack that it select
- calamares needs to be able to put the luks2 parts together; that is something it probably does **not** do right now, because of the above^^ (although there is code that checks for / does things with Luks2; it requires KPMCore 4 which most distro's have now)
- The UI needs to be adjusted to pick one or the other (or maybe just "switch to luks2 by default" -- once the moving parts are in place, the selection is not interesting since, as you put it, there's not much reason to stick with outdated technology)

**Comment by Bleuzen** [comment]
@adriaandegroot 
> luks2 needs to work **at all** [...]

It does. I have it in use on many systems now for about two years (since it became the default). Can't report any problems with it.

Only thing that doesn't work yet is the Argon2i key derival function in Grub. But only in grub, so this only becomes a problem if /boot is encrypted with luks2. (Therefore the warning message I suggestet in the main post.)

For everything else (see the use-cases / examples in the main post) luks2 works perfectly fine.

> [...] it's not clear that all the parts are in place to do that yet; this will also depend on the distro and the other parts of that stack that it select

Which parts do you mean that could be missing?

From distro level it should be safe now to implement luks2. Even Ubuntu 18.04 supports it. Only users with very old distros could get problems (something with packages from pre December 2017).
But those can continue to use luks1 perfectly fine since I'm talking about an optional option, not forcing luks2 on anyone all of the sudden.

Also the chance that very old distros would use the very newest version of Calamares is low.
And the chance that exactly the users on these old distros would select the "use luks2" checkbox in manual partitioning mode is even lower.

While I think this is negligible, if you really want, add something like this to the warning message I suggested in the main post:
"This does not work on older distributions with cryptsetup < 2.0.0 (before december 2017)!"

> calamares needs to be able to put the luks2 parts together

From what I know, to use luks2 is as simple as passing `--type luks2` to `cryptsetup` when creating the partition.

Is there something special about Calamares that makes this more complicated?

> (although there is code that checks for / does things with Luks2; it requires KPMCore 4 which most distro's have now)

Good :)

> The UI needs to be adjusted to pick one or the other [...]

... for example with a simple checkbox, so changes are small in this area (see the screenshot in the main post)


> (or maybe just "switch to luks2 by default" -- once the moving parts are in place, the selection is not interesting since, as you put it, there's not much reason to stick with outdated technology)

I think switchting to luks2 by default is not a good idea just yet. Maybe do it when grub supports Argon2i. Or auto detect and select, something like:

```
if (cryptsetup.version > "2.0.0" && no "/boot" on partition) {
    use luks2
} else {
    use luks1
}
```

However this is something that could be done in the future.

---

The point of this feature request is to get something to work with (an option to use luks2) as soon as possible while making it optional and switching over by default later when fully ready.

**Comment by stikonas** [comment]
> There are three parts to this request:
> 
>     * luks2 needs to work **at all**; the issue Phil refers to discusses that and it's not clear that all the parts are in place to do that yet; this will also depend on the distro and the other parts of that stack that it select
> 
>     * calamares needs to be able to put the luks2 parts together; that is something it probably does **not** do right now, because of the above^^ (although there is code that checks for / does things with Luks2; it requires KPMCore 4 which most distro's have now)
> 
>     * The UI needs to be adjusted to pick one or the other (or maybe just "switch to luks2 by default" -- once the moving parts are in place, the selection is not interesting since, as you put it, there's not much reason to stick with outdated technology)

I am kind of waiting for all pieces to be in place (Grub in particular) before switching default in Partition Manager too.

**Comment by abucodonosor** [comment]
@Bleuzen 

>It does. I have it in use on many systems now for about two years (since it became the default). Can't report any problems with it.

We can talk about what is working and what not but not the point...

>Only thing that doesn't work yet is the Argon2i key derival function in Grub. But only in grub, so this only becomes a problem if /boot is encrypted with luks2. (Therefore the warning message I suggestet in the main post.)

Which is the calamares *default*. And no there is no release of grub2 with any Luks2 support by now, so all this works but ... arguments are useless until there is a grub2 released with that support, not random distros patches and backports.

**Comment by Bleuzen** [comment]
@abucodonosor 
> [...] And no there is no release of grub2 with any Luks2 support by now, so all this works but ... arguments are useless until there is a grub2 released with that support, not random distros patches and backports.

You are right on this one, I should have looked up the official grub releases.

Anyway, the thing with grub does not play a role here as this is not about using luks2 everywhere but it is only an optional thing even with a warning. So this does not stop what I described in the main post from being implemented.

**Comment by edmundlaugasson** [comment]
Would it be possible to allow choose LUKS1 for /boot and LUKS2 for other partitions? This could be useful even we finally reach full GRUB2 support. People might need to choose different LUKS versions for different partitions. In addition, there could be useful to allow to bind multiple disks into one with [LVM](https://en.wikipedia.org/wiki/Logical_Volume_Manager_%28Linux%29), before starting installation with encryption. Sometimes people have multiple disks, that should be binded into one virtually.

**Comment by philmmanjaro** [comment]
Sounds like special cases most of us don't use. However feel free @edmundlaugasson to do a PR for review ...

**Comment by edmundlaugasson** [comment]
Wouldn't say so. Many computers have more than one storage device, so it is quite common case actually... Don't have much time for PR - I'm reflecting user and sysadmin view, not a developer. If you wan't more usability and less hard hacking at CLI, then you are welcome to consider my ideas...

**Comment by edmundlaugasson** [comment]
I would suggest to add multiple choices:
- use LUKS1
- use LUKS2
In future, when LUKS3, etc comes out, it would be easier to add more choices, pull-down menu, similar checking like currently boot flags, etc.
Looks like [GRUB2 supports LUKS2](https://git.savannah.gnu.org/cgit/grub.git/commit/?id=365e0cc3e7e44151c14dd29514c2f870b49f9755) since January 10, 2020.
There is an article [GRUB Boot Loader Adds Support For LUKS2 Encrypted Disks - Phoronix](https://www.phoronix.com/scan.php?page=news_item&px=GRUB-Boots-LUKS2-Disk-Encrypt) from January 10, 2020.
I really hope, that current feature request will be implemented. This gives flexibility regardless of what GRUB does support. Sysadmins know it, will choose appropriate LUKS version and the world will be more secure and private.

**Comment by kkofler** [comment]
GRUB supports LUKS2 only with the legacy crypto ciphers from LUKS1, not with the new Argon2i ciphers that are the default in LUKS2. So you cannot just create a LUKS2 partition and have it work in GRUB, you have to set special options, and the result is not really any more secure than LUKS1.

**Comment by edmundlaugasson** [comment]
Therefore I proposed to allow /boot to be installed as LUKS1 and rest of partition as LUKS2. This cipher issue is mentioned also on that article I mentioned. But currently we cannot choose LUKS version graphically, this is why current issue was born. I also mentioned, that we just need ability to choose LUKS version in Calamares and that's it. This is actually not the issue about GRUB support to certain LUKS version. I only mentioned these articles, because the ability of choosing LUKS version is important during installation.

**Comment by IstotaSwiatla** [comment]
> Sounds like special cases most of us don't use. However feel free @edmundlaugasson to do a PR for review ...

Apart from kids playing games, I can't believe anyone who is using Linux as a daily driver for professional work would take the risk of installing any OS without encryption, particularly on a laptop that can be lost or stolen. And I can't believe anyone else (well, apart from @edmundlaugasson) does not see the need for sensible solution... I guess most people simply do not care, they want to just click the button and get stuff installed and working. And if that's not the case, they go back to MacOS or Windows.

Lack of LUKS2 support is exactly the reason I moved back from Manjaro to Debian - Debian is the only distro I know of where the installer is really flexible, in particular it supports simple option of having unencrypted /boot and LUKS2 encrypted root partition (without LVM). Manjaro used to have Architect ISO which gave more or less same flexibility during installation, but unfortunately is no longer maintained, and we have to stick to Calamares.

Before I get bombarded with hate, yes I know I can read up and install Arch if I want real flexibility, but I simply cannot invest that much time in the process.

**Comment by edmundlaugasson** [comment]
What?! I am one of the strong supporter of encrypted installation, especially with LUKS2 and you (@IstotaSwiatla ) are writing so wrongly?! Did you (@IstotaSwiatla) read, what I wrote? Sounds like not. Did you (@IstotaSwiatla) make a typo with username? Besides, I have manually installed Linux Mint with LVM and LUKS2 ca year ago Manually, because there is no full disk encryption GUI installer like Calamares. Just /boot is on separate partition with LUKS1, rest are encrypted with LUKS2. Currently considering to do same with Endeavour OS.

**Comment by stikonas** [comment]
In any case it's not a matter of anybody opposing LUKS2. It's more of a matter of work required vs benefits. Waiting until everybody down the line properly supports LUKS2 and then switching the default is 5 minutes of work. Implementing an option to pick LUKS version is significantly more work. In the longer term we'll end up with LUKS2 as default in both cases, so we have no long term benefits, at best short term benefits. But even then, what are the benefits of LUKS2 that you are specifically looking at? Disk format changes might be helpful for cryptsetup maintainers but users wouldn't gain additional security. The main difference is default key derivation function (which only matters if you use short passphrase with insufficient entropy).

And for the short term... Is anybody willing to do that work? If yes then patches are welcome.
Personally I can just wait until next GRUB version (I think 2.11) and then we simply switch default to LUKS2.

**Comment by philmmanjaro** [comment]
Right, then it should have been called **initial support** or some. Well, at least there is progress and a reason why **luks1** is still the de-facto and save default setting for now.

**Comment by IstotaSwiatla** [comment]
> What?! I am one of the strong supporter of encrypted installation, especially with LUKS2 and you (@IstotaSwiatla ) are writing so wrongly?! Did you (@IstotaSwiatla) read, what I wrote? Sounds like not. 

Either you misread my post, or I didn't express myself clearly - I'm not a native English speaker. I said I can't believe nobody cares about LUKS2, and you seem to be the only person who does.

And yes I did read your post, and your post was the reason I added mine - to support your view and the case for adding LUKS2 support to Calamares. There is really no need to go ballistic and vent your anger on me, you got the wrong end of the stick.

**Comment by stikonas** [comment]
> I guess encryption should be at least one (better 2-3 in quantum computing era like currently) step further of current cyber criminals.

Disk encryption is generally already quantum resistant. It uses symmetric encryption and quantum-computers are only marginally helpful with brute-force but that's not enough to break AES-256.

**Comment by edmundlaugasson** [comment]
I referred to scientists research from. 2017, that said 6681 qubits would be needed to break AES-256. Also mentioned, that this might be reached unexpectedly sooner than we imagine. Don't know where that million qubits comes. I rely on scientists research.

**Comment by adriaandegroot** [comment]
Hi. Please use other communications media -- forums, email, TikTok videos -- to discuss research on quantum computing and the state of the art. None of the discussion, however interesting, helps move the issue forward. LUKS2 is interesting if someone thinks it is. The best way to move it forward is to submit a PR for it (across however many repositories need changing, then) and convincing a distro to ship it.

**Comment by AKmatiAK** [comment]
I need luks2 too, because I have 4Kn HDD and luks1 doesn't support 4096 byte sector size. It works with 512b, yet downgrades performance. Doesn't matter PBKDF2 or Argon2, luks1 is also PBKDF2 so by switching to luks2 there are only profits and it's compatible with GRUB now, or can be used with systemd-boot instead for Argon2. In my opinion staying with luks1 no longer makes sense.

**Comment by S0PEX** [comment]
+1
Is there any way to remove the `--type luks1` parameter from calamares?
I would rather run grub with `luks2` and then just replace my slot with `argon2id` once supported in (2.11) instead of converting my luks later on.
Alternatively we could just use LILO as 2nd stage with grub and then use `luks2 + argon2` directly.

**Comment by adriaandegroot** [comment]
`--type luks1` is in KPMCore, not in Calamares. Calamares passes enum values to KPMCore, as you've already spotted. Using `FileSystem::Luks2` there would do the trick. A PR would be welcome. I'd suggest something along these lines:

- [ ] add a data member to the `Config` class in the *partition* module. I'd suggest calling it `m_LUKSGeneration` or something like that, and making it an enum (enum class, really) so it's easy to update and use in a type-safe fashion.
- [ ] set that data member from `setConfigurationMap()` to a default value.
- [ ] add a key to `partition.conf` and the schema for that file, e.g. `luksGeneration` or `luks_generation`, give it string values from an enumeration (e.g. "luks", "luks1", "luks2")
- [ ] add a `NamedEnum` for those names, mapping the names to the values of the enum class. You can find an example in the `InstallChoice` enum and corresponding code.
- [ ] set the data member with the name obtained from the config file, using the enum-mapping.
- [ ] modify `createNewEncryptedPartition()` appropriately. It might be possible to pass in Luks or Luks2 via the partition roles, otherwise there might need to be a new parameter added to the call. Check with the `Config` object before calling the modified function, to decide which kind of Luks to pass in.

**Comment by S0PEX** [comment]
@adriaandegroot 
Good evening,

I made the needed changes (https://github.com/calamares/calamares/pull/1964) to the `Config.cpp`, but I ran into the problem that I don't really know how to pass the luks generation value to the create function.

As soon as I have time I will take a closer look at the project and figure out how these roles are set and check If I can use them to pass the luks generation.

**Comment by jtheoof** [comment]
You can have a look at #2047 and #2070. Luks2 is now supported if maintainers activate it.

**Comment by kkofler** [comment]
#2047 allows the distribution packager to select the LUKS version to use, but does not add UI controls to allow the user to override the choice, as this feature request is asking for. And #2070 is just an internal code cleanup and does not add anything user-visible at all.

**Comment by AKmatiAK** [comment]
Plz make this switch someone, today I tried installing with luks2 encryption manually, but calamares crashes then on manjaro install. How loooong I must stay with this shitty windows and veracrypt :(

**Comment by AKmatiAK** [comment]
today I digged a bit about customizing calamares by configuration files, and I found that new PR introduced ability to specify luks version in /src/calamares and somewhere is partition.conf. But we need to wait for new release :c then you can simply change luks1 to luks2 in text file, that's pretty easy
